### ISSUE TITLE
Support passwordless login as an option

### ISSUE THREAD

**Issue by henrjk** [anvilresearch/connect#124]
Medium introduced a passwordless signin/signup. Here is a  [related blog post](https://medium.com/the-story/signing-in-to-medium-by-email-aacc21134fcd).

Passwordless login only requires entry of an email. However to validate one has to also validates a link send by email. This is similar to how password reset systems typically work.  Of course one could also send the link over another channel in particular a text message (SMS).

It should not be sufficient to simple enter an email address to login on a different device. Instead a new signin link should be required.

**Comment by christiansmith** [comment]
Very interesting approach to authentication. Thanks for posting this. I think we should add this type of behavior to Connect as an option. Who wants to work on it?

**Comment by henrjk** [comment]
##Here is what medium's currently does##

Sign up and sign in is not differentiated in the UI. 
 
![medium_signin_signup](https://cloud.githubusercontent.com/assets/7439282/8786078/6e591b62-2f2c-11e5-80f6-7b8a07f21d01.png)

After selecting sign in/sign up the following pages are shown:

![medium_signinup_flow](https://cloud.githubusercontent.com/assets/7439282/8786221/16ae4012-2f2d-11e5-87e6-3f89a2604d1a.png)

Without an account created one gets an email: with text like this which also has a link in it:

*Click and confirm that you want to create an account on Medium. This link will expire in fifteen minutes and can only be used once.*

When you visit the link, the browser asks a few pages more about account creation:the first page allows one to change the undesirable defaults (which are just the email address) for username and Full name and also has legal links. The next pages are about interests and who to follow to customize the medium experience. I makes good sense to me to only having to do this once the email validation actually worked. 
Then one is logged in. I remained logged in until I explicitly log out. 

There may also be a time when the login expires. Once logged out, one has to go through the original flow, except that the emailed link now simply logs in, instead of asking further account creation questions. The email is also worded differently to make this clear. The login links also expire in 15 minutes.

**Comment by vsimonian** [comment]
I think we can get this working really easily once #123 is filled out. We're basically just issuing an expiring token for authentication and e-mailing it.

**Comment by christiansmith** [comment]
Thinking we probably want to create a separate [protocol](https://github.com/anvilresearch/connect/tree/master/protocols) for this. That way it gets enabled/configured just like everything else.

**Comment by vsimonian** [comment]
I agree it should be a separate protocol/provider; I just think we can repurpose whatever we develop now for it.

**Comment by topperge** [comment]
This should be looked at for reference:
https://github.com/florianheinemann/passwordless

**Comment by christiansmith** [comment]
Great link @topperge. Thanks!

**Comment by henrjk** [comment]
I looked a bit at the sources florianheinemann/passwordless @topperge pointed to. Here are some notes about this:

1. passwordless is not based on [passport](https://github.com/jaredhanson/passport). It has one small self contained direct dependencies `bs58` declared in package.json but it does requires a [tokenstore](https://github.com/florianheinemann/passwordless-tokenstore/blob/master/lib/tokenstore.js) for which there is an abstract class.
2. The tokenstore stores the outstanding tokens, which can activate the associated accounts. 
3. `passwordless.requestToken` determines the validation link, stores it in the store and then delivers it.
4. `passwordless.acceptToken` validates a link and authenticates the user (`req.user`). It also removes the token from the store unless it is configured to allow reuse.
5. Other passwordless functions are restricted and logout as well as setup related functions init, addDelivery, sessionSupport.
6. The delivery methods can be configured. The **[simple-mail example]**(https://github.com/florianheinemann/passwordless/blob/master/examples/simple-mail/app.js) shows how to use email delivery using email.js. In addition the example uses a [MongoDb based implementation for the tokenstore](https://github.com/florianheinemann/passwordless-mongostore). 

I found an Angular example app for utilizing [passwordless with a stateless API server](https://github.com/billstron/passwordless-stateless-angular) which uses an in memory token store. It seems one could do away with storing tokens in memory entirely and instead use crypto to ensure that the tokens were produced by the auth server and have not been changed. But my impulse would be that a database is not needed to store outstanding tokens as their number should be limited as they expire relatively quickly.

Some questions:

1. Currently all protocols are based on passport. Is it a concern regarding using this in connect that florianheinemann/passwordless is not based on passport?

2. In case one uses a memory based token store would one use bcrypt? I am thinking no but the passwordless-stateless-angular example uses bcrypt hashing nonetheless.

3. The biggest question is how does passwordless auth fit into connect?

When passwordless.requestToken creates the link one could internally create a user but mark it somehow has requiring a 'link activation' before it can be used. This would be a similar status as an account with an expired password might be managed.  
How would the communication with the client work?

Does the requestToken flow fits in the standard open id connect flows somehow?  
How does password reset or link activation fit into the standards?

**Comment by henrjk** [comment]
I have been working on this and it should now be ready for review:

The configuration and UI is described at https://github.com/henrjk/connect-docs/blob/pwless/providers/Passwordless.md.

I also created a PR for the code. Since this is rather sizeable I will add some more comments hopefully today, but it could be Monday, depending on other commitments.
